### PR TITLE
Hermitian, ishermitian now check diagonals

### LINKED ISSUE #196: Bug with Hermitian() and ishermitian()
Hello,
It seems that the methods Hermitian() and ishermitian() are not working properly.

```
test = Hermitian(rand(4,4)+im*rand(4,4))
```

generates a matrix that is not hermitian: the diagonal elements are complex.

Also, trying

```
ishermitian(test)
```

returns true.

### LINKED ISSUE #10671: Bug with Hermitian() and ishermitian()
Hello,
It seems that the methods Hermitian() and ishermitian() are not working properly.

```
test = Hermitian(rand(4,4)+im*rand(4,4))
```

generates a matrix that is not hermitian: the diagonal elements are complex.

Also, trying

```
ishermitian(test)
```

returns true.

### PR DESCRIPTION
**Pull request by jiahao**
Fix JuliaLang/LinearAlgebra.jl#196

I also had to fix some tests which had incorrectly neglected to check the diagonals.

### CODE REVIEW

File: base/linalg/symmetric.jl [hunk 1]
```diff
@@ -8,7 +8,13 @@ immutable Hermitian{T,S<:AbstractMatrix} <: AbstractMatrix{T}
     data::S
     uplo::Char
 end
-Hermitian(A::AbstractMatrix, uplo::Symbol=:U) = (chksquare(A);Hermitian{eltype(A),typeof(A)}(A, char_uplo(uplo)))
+function Hermitian(A::AbstractMatrix, uplo::Symbol=:U)
+    chksquare(A)
+    all(isreal(diag(A))) || throw(ArgumentError(
```
Comment (andreasnoack) [inline base/linalg/symmetric.jl]: I think we should do this without allocation.

File: base/linalg/symmetric.jl [hunk 2]
```diff
@@ -26,7 +32,7 @@ convert{T,S<:AbstractMatrix}(::Type{Hermitian{T,S}},A::Hermitian) = Hermitian{T,
 convert{T}(::Type{AbstractMatrix{T}}, A::Hermitian) = Hermitian(convert(AbstractMatrix{T}, A.data), symbol(A.uplo))
 copy{T,S}(A::Symmetric{T,S}) = Symmetric{T,S}(copy(A.data),A.uplo)
 copy{T,S}(A::Hermitian{T,S}) = Hermitian{T,S}(copy(A.data),A.uplo)
-ishermitian(A::Hermitian) = true
+ishermitian(A::Hermitian) = all(isreal(diag(A.data)))
```
Comment (andreasnoack) [inline base/linalg/symmetric.jl]: When we have made the check in the constructor it shouldn't be necessary here.

File: test/linalg/symmetric.jl [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+# Hermitian matrix exponential
+let A1 = randn(4,4) + im*randn(4,4)
+    A2 = A1 + A1'
+    @test_approx_eq expm(A2) expm(Hermitian(A2))
+end
+
+let n=10
+    areal = randn(n,n)/2
+    aimg  = randn(n,n)/2
+    breal = randn(n,2)/2
```
Comment (andreasnoack) [inline test/linalg/symmetric.jl]: I don't think `b` is used here anymore

File: test/linalg/symmetric.jl [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    bimg  = randn(n,2)/2
+    debug && println("symmetric eigendecomposition")
+    for eltya in (Float32, Float64, Complex64, Complex128, BigFloat, Int)
+        a = eltya == Int ? rand(1:7, n, n) : convert(Matrix{eltya}, eltya <: Complex ? complex(areal, aimg) : areal)
+        asym = a'+a                 # symmetric indefinite
+        ε = εa = eps(abs(float(one(eltya))))
+
+        for eltyb in (Float32, Float64, Complex64, Complex128, Int)
```
Comment (andreasnoack) [inline test/linalg/symmetric.jl]: ...so this loop can probably be removed

File: base/linalg/symmetric.jl (hunk 2 above)
Comment (jiahao) [inline base/linalg/symmetric.jl]: This is a separate function from the constructor

Comment (andreasnoack) [inline base/linalg/symmetric.jl]: but if a real diagonal is enforced during construction it is okay to assume that it is real at this point.

Comment (jiahao) [inline base/linalg/symmetric.jl]: yes that makes sense.

**Comment by jiahao** [comment]
Can anyone shed light on the [32-bit build failure](https://travis-ci.org/JuliaLang/julia/jobs/56335725#L3957)?

```julia-debug: ast.c:686: jl_lam_vinfo: Assertion `(((jl_value_t*)(((jl_taggedvalue_t *) ((char *)((l)) - __builtin_offsetof (jl_taggedvalue_t, value)))->type_bits&~(size_t)3))==(jl_value_t*)(jl_expr_type))' failed.```

**Comment by jiahao** [comment]
Build failure seems unrelated.

**Comment by tkelman** [comment]
This is causing intermittent failures, the tolerances look a little too tight

**Comment by jiahao** [comment]
Could you point me to a failure?

**Comment by tkelman** [comment]
Here's 4, haven't even checked appveyor.
https://travis-ci.org/JuliaLang/julia/jobs/57350258
https://travis-ci.org/JuliaLang/julia/jobs/57350849
https://travis-ci.org/JuliaLang/julia/jobs/57354524
https://travis-ci.org/JuliaLang/julia/jobs/57354525

**Comment by tkelman** [comment]
Bump. Any suggestions for mitigating these?

**Comment by jiahao** [comment]
I wrote a simple PR but I got [an inscrutable error](https://travis-ci.org/JuliaLang/julia/jobs/57437534#L3839).

**Comment by andreasnoack** [comment]
I think we should revert because it gives some annoying exceptions this are often unwanted. We don't check that a matrix is symmetric in `cholfact` so I don't think we should check the input of `Hermitian`/`Symmetric`. We should simply ignore the imaginary part of the diagonal and document the behavior.

**Comment by tkelman** [comment]
Ignoring the diagonal sounds like a pretty bad idea - the constructors here should verify the invariants that their types represent.

**Comment by andreasnoack** [comment]
No. The point is that the user already knows that the matrix has the property or sohuld be interpreted in that way. Similarly to allowing
```jl
julia> LowerTriangular(randn(3,3))
3x3 LowerTriangular{Float64,Array{Float64,2}}:
 -0.193406   0.0       0.0    
 -0.225881  -1.29177   0.0    
  1.83933    1.77965  -1.24135
```

**Comment by tkelman** [comment]
Unlike `LowerTriangular`, you aren't proposing ignoring uninitialized data - you're proposing ignoring actual imaginary parts of values of elements that matter. You'd have to zero them and make the `Hermitian` constructor mutating, which is surprising and inconsistent with the rest of the wrapper types.

**Comment by andreasnoack** [comment]
It's not necessary to mutate anything. The methods for `Hermitian`/`Symmetric` should just ignore any imaginary components in the diagonal. That is also that LAPACK is doing and what we are doing for Cholesky
```
julia> cholfact(fill(complex(1.0,1.0), 1, 1))
Base.LinAlg.Cholesky{Complex{Float64},Array{Complex{Float64},2}} with factor:
1x1 UpperTriangular{Complex{Float64},Array{Complex{Float64},2}}:
 1.0+0.0im
```

**Comment by tkelman** [comment]
Every single method then. getindex and everything else you might ever call on a `Hermitian` will need extra special casing to ignore the imaginary component any time it touches a diagonal. Sounds very very easy to get wrong for not much benefit, easier to explicitly verify up front.

**Comment by andreasnoack** [comment]
No. It's not easy to verify without a lot of unnecesary exceptions. One of the main reasons to use `Hermitian` is when floating point errors results in a computation like `A = QΛQ'` begin slightly non-hermitian. Then you want to use `Hermaitan(A)`, but it defeats the purpose to throw an exception in that case.

I dont' think ingoring imaginary parts in the diagonal will be that demanding. Making the very small change to `getindex` will probably take care of most of it.

**Comment by tkelman** [comment]
`tril` and friends as well, probably others. And when you send the data to a C library?

**Comment by andreasnoack** [comment]
There might be a few other places in `symmetric.jl`, but the file is only 239 lines so it's very doable. As far as I can tell, we are calling out to three libraries with `Hermitian`: `BLAS`, `LAPACK` and `CHOLMOD`. The first two use the assumption I advocate for here and `CHOLMOD` checks the imaginary part like we are doing now.

**Comment by tkelman** [comment]
Methods that use in any way other than getindex, or leave alone without checking, diagonal elements for any abstract supertype of Hermitian would need to be made aware of this special case, or have a more specific method written for Hermitian. I'm anticipating there will be quite a few places in base that would get this wrong. This is the kind of thing where we start needing type-based coverage stats.

I'm more referring to users sending their data to C libraries outside of base. The proposal is to store undefined data that we intentionally want to be meaningless inline within the same diagonal scalar values. This could be documented and tested thoroughly for corner cases, and left up to the user to manually zero the imaginary components for this ccall case, but this isn't something we've done any other places to my knowledge.

**Pull request merged**